### PR TITLE
Remove flare tag from title and hightlight it within the tags row

### LINKED ISSUE #10523: Adjust flare tag display on Feed card
## Problem
Right now, the "flare tag" on feed cards sit above the title, but this visual design is clunky and not what we want to keep in long-term.

## Proposed solution
Change the way we display flare tag by highlighting it within the row of tags below the title. 

[Design mockup](https://www.figma.com/file/lMIzNY8ORLej9XInAqx34w/Untitled?node-id=0%3A1)

### PR DESCRIPTION
**Pull request by akshayymahajan**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description
This PR does 2 things:
1. Remove the Flare Tag from the post title.
2. Moves the flare tag to the first position within the list of tags and highlights it.

## Related Tickets & Documents
Closes #10523

## QA Instructions, Screenshots, Recordings
image

## Added tests?

- [ ] yes
- [x] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

File: app/javascript/articles/components/TagList.jsx [hunk 1]
```diff
@@ -1,16 +1,35 @@
...
+      {flare_tag && (
+        <a className="crayons-tag" href={`/t/${flare_tag.name}`}>
+          <span
+            className="crayons-story__flare-tag"
+            style={{
+              background: flare_tag.bg_color_hex,
+              color: flare_tag.text_color_hex,
+            }}
+          >
+            {`#${flare_tag.name}`}
```
Comment (ludwiczakpawel) [inline app/javascript/articles/components/TagList.jsx:20]: Can you wrap `#` with `<span className="crayons-tag__prefix">` similar to [how we do it with other tags](https://github.com/forem/forem/pull/10528/files#diff-bcb8e7ee1cc3e3133c3796a5a01689f0R26)?

**Review by ludwiczakpawel — CHANGES_REQUESTED** [review]
Oh damn, I love it! Thank you so much! I have one very tiny change request - other than that it's awesome!

File: app/javascript/articles/components/TagList.jsx (hunk 1 above)
Comment (akshayymahajan) [reply app/javascript/articles/components/TagList.jsx:20]: Sure thing!

**Comment by ludwiczakpawel** [comment]
I've just realized, that when you scroll down quite a lot on your feed we actually load feed cards differently (I don't know why...) - so the feed card code is generated in multiple places. That jsx component is one of them and it works great. But we gonna need it in other places too.. Do you think you can look at it as well? You can also spot it in logged out version of the page where we also load feed card differently.

**Comment by akshayymahajan** [comment]
Sure.. I'll look into it..

**Comment by ludwiczakpawel** [comment]
@akshayymahajan awesome! I'm super excited about this one  :)

Btw. here are some references you may find useful: 

- [_single-story.html.erb](https://github.com/forem/forem/blob/master/app/views/articles/_single_story.html.erb)
- [buildArticleHTML.js](https://github.com/forem/forem/blob/master/app/assets/javascripts/utilities/buildArticleHTML.js)

**Comment by akshayymahajan** [comment]
Hey @ludwiczakpawel, I have updated both files and tested in the logged out page and the search results. Can you check again?

File: app/assets/javascripts/utilities/buildArticleHTML.js [hunk 2]
```diff
@@ -25,8 +25,45 @@ function buildArticleHTML(article) {
...
+        article.flare_tag.name +
+        "' class='crayons-tag'>" +
+        "<span class='crayons-story__flare-tag' style='background:" +
+        article.flare_tag.bg_color_hex +
+        ';color:' +
+        article.flare_tag.text_color_hex +
+        "'><span className='crayons-tag__prefix'>#</span>" +
+        article.flare_tag.name +
+        '</span>' +
+        '</a>';
```
Comment (ludwiczakpawel) [inline app/assets/javascripts/utilities/buildArticleHTML.js:46]: We should drop this extra container `crayons-story__flare-tag` - it's not needed anymore.

```suggestion
    flareTag = `<a href="/t/${article.flare_tag.name}" class="crayons-tag" style="background: ${article.flare_tag.bg_color_hex}; color: ${article.flare_tag.text_color_hex};">
      <span class="crayons-tag__prefix>#</span>
      ${article.flare_tag.name}
    </a>`;
```

File: app/assets/stylesheets/components/stories.scss [hunk 3]
```diff
@@ -135,7 +135,7 @@
 
   &__flare-tag {
```
Comment (ludwiczakpawel) [inline app/assets/stylesheets/components/stories.scss:136]: We should now be able to remove this entire class, right?

File: app/javascript/articles/components/TagList.jsx [hunk 4]
```diff
@@ -1,16 +1,36 @@
...
+  let tagsToDisplay = tags;
+  if (flare_tag) {
+    tagsToDisplay = tagsToDisplay.filter((tag) => tag !== flare_tag.name);
+  }
+  return (
+    <div className="crayons-story__tags">
+      {flare_tag && (
+        <a className="crayons-tag" href={`/t/${flare_tag.name}`}>
+          <span
+            className="crayons-story__flare-tag"
```
Comment (ludwiczakpawel) [inline app/javascript/articles/components/TagList.jsx:14]: Similarly to what I suggested in [previous comment](https://github.com/forem/forem/pull/10528/files#diff-419ded8bb2b41fca12af2061c1272094), we should drop `.crayons-stroy__flare-tag`

File: app/assets/javascripts/utilities/buildArticleHTML.js (hunk 2 above)
Comment (ludwiczakpawel) [reply app/assets/javascripts/utilities/buildArticleHTML.js:46]: The flare tag is basically now a classic `crayons-tag` with custom background and color.

File: app/views/articles/_single_story.html.erb [hunk 5]
```diff
@@ -51,11 +51,19 @@
...
         <a href="<%= story.path %>" data-preload-image="<%= cloud_cover_url(story.main_image) %>" id="article-link-<%= story.id %>">
-          <%= render "articles/tag_identifier", story: story, tag: @tag %>
+          <%= story.title %>
         </a>
       </h2>
       <div class="crayons-story__tags">
-        <% story.cached_tag_list_array.each do |tag| %>
+        <% flare_tag = FlareTag.new(story, @tag).tag %>
+        <% if flare_tag %>
+          <a href="/t/<%= flare_tag.name %>" class="crayons-tag"><span class="crayons-story__flare-tag" style="background:<%= flare_tag.bg_color_hex %>;color:<%= flare_tag.text_color_hex %>"><span className="crayons-tag__prefix">#</span><%= flare_tag.name %></span></a>
```
Comment (ludwiczakpawel) [inline app/views/articles/_single_story.html.erb:60]: And same here.

**Review by ludwiczakpawel — CHANGES_REQUESTED** [review]
Awesome! I left few last change request - nothing really major :)

**Comment by akshayymahajan** [comment]
@ludwiczakpawel updated!

File: app/assets/stylesheets/components/stories.scss [hunk 6]
```diff
@@ -133,21 +133,6 @@
...
-    position: relative;
-    top: -2px;
-    line-height: 1em;
-    display: inline-block;
-    border-radius: var(--radius);
-    background: var(--base-60);
-    margin-right: var(--su-1);
-    color: var(--base-inverted);
-  }
-
```
Comment (ludwiczakpawel) [inline app/assets/stylesheets/components/stories.scss:150]: Whoops, looks like this is still needed for search results. I apologize for the confusion! Other than that, it's good to go 👍

**Review by ludwiczakpawel — APPROVED** [review]
One tiny change and it's ready!

**Review by lisasy — APPROVED** [review]
Nice! Thank you :) This looks great.

File: app/assets/stylesheets/components/stories.scss (hunk 6 above)
Comment (akshayymahajan) [reply app/assets/stylesheets/components/stories.scss]: Added back!

**Comment by rhymes** [comment]
@akshayymahajan it looks like some of the tests are failing, please take a look at the build here: https://github.com/forem/forem/pull/10528/checks?check_run_id=1201884642

**Comment by akshayymahajan** [comment]
Hey @rhymes, I have updated the code to fix the failing test.

**Review by ludwiczakpawel — CHANGES_REQUESTED** [review]
Hey, I've taken it for another ride and there's still something a little bit off, when I scroll down the feed. Here's a [video](https://d.pr/v/Pzx0DE) showing the issue. Looks like on the feed cards that are being loaded with infinity scroll we don't apply appropriate colors/styling to the flare tag.

**Comment by akshayymahajan** [comment]
Hey @ludwiczakpawel, looks like the data being returned from the API has some null values. Not sure if this is related to the changes I made. Would really appreciate some help.

Screenshot 2020-10-05 at 2 55 34 PM

**Comment by ludwiczakpawel** [comment]
Interesting. I can see this working properly on prod dev.to... 

@rhymes I SUMMON YOU.

summon rhymes

**Comment by rhymes** [comment]
Hi @akshayymahajan, that indicates that the tag has no defined colors: 

https://github.com/forem/forem/blob/a943f045c56237f4ec8a4b2e504a6ac3f7fb0db9/app/models/tag.rb#L26-L27

both hex colors can be `nil`

**Comment by akshayymahajan** [comment]
Thanks, @rhymes. @ludwiczakpawel in that case, I think it would be best to add back the `span.crayons-stroy__flare-tag` wrapper in all the places. What do you say?

**Comment by ludwiczakpawel** [comment]
@rhymes are you sure? can you please checkout the screen recording here: https://d.pr/v/Pzx0DE - it's the same tag. one time it has colros applied and the other time does not... which means that this tag in general has colors applied but they are not returned in certain cases.

**Comment by rhymes** [comment]
@ludwiczakpawel I'm not very familiar with how flare tags work but I came up with these two clues:

`article.flare_tag` always calls the `FlareTag` class with just the article object, without the "except tag" parameter:

https://github.com/forem/forem/blob/a943f045c56237f4ec8a4b2e504a6ac3f7fb0db9/app/models/article.rb#L312-L314

https://github.com/forem/forem/blob/a943f045c56237f4ec8a4b2e504a6ac3f7fb0db9/app/services/flare_tag.rb#L18-L21

It looks to me that if `FlareTag` were to be initialized with a second argument, it would send back `nil` even if the `Tag` itself has the colors assigned to it:

https://github.com/forem/forem/blob/a943f045c56237f4ec8a4b2e504a6ac3f7fb0db9/app/services/flare_tag.rb#L37-L40

The clue for what might be going on in your recording is to be found in:

https://github.com/forem/forem/blob/a943f045c56237f4ec8a4b2e504a6ac3f7fb0db9/app/views/articles/_tag_identifier.html.erb#L1-L3

which is passing `@tag` as the "except tag". 

This comes from https://github.com/forem/forem/blob/master/app/views/articles/_single_story.html.erb

I seem to recall that the first page of the feed is rendered one way, subsequent pages are rendered in a different way but I'm not sure how all that works, @joshpuetz or @nickytonline might instantly know what's going on here.

There might be a different behavior between the flare tag appearing in the first batch of articles and the following batches

**Comment by akshayymahajan** [comment]
I tried removing the "except tag" from the call, but it still returns `null` value for colors. I also tried the latest master branch locally and the JSON returned still has `null` values. Looks to me like this is something related to the local development environment.

**Comment by ludwiczakpawel** [comment]
@akshayymahajan let me give it another look tomorrow and I'll get back to you then.

**Comment by ludwiczakpawel** [comment]
Yea, looks like it's being broken on master as well.. So I'll go ahead and approve this PR :) Thanks for all the work you've done, it's really something I've been looking for for a quite long time so I appreciate your effort :)

**Review by ludwiczakpawel — APPROVED** [review]

**Pull request merged**